lib/task.js

```diff
@@ -0,0 +1,401 @@
+      let target = targets[i];
+
+      for (let name in props) {
+        let propInfo = props[name];
+
+        let keys = _createKeys(target[name], propInfo.keys, propInfo.type, opts);
+
+        let type = propInfo.type;
+
+        tracks.push({
+          target: target,
+          property: name,
+          type: type,
+          keys: keys,
+          duration: keys[keys.length - 1].endTime,
+          delay: keys[0].delay,
+          lerp: getLerpFunc(type),
+        });
+      }
+    }
+  }
+
+  return tracks;
+}
+
+export default class Task {
+  constructor(engine, targets, props, opts) {
+    this._next = null;
+    this._prev = null;
+    this._engine = engine;
+
+    // animation options
+    this.loop = 1;
+    if (opts.loop !== undefined) {
+      this.loop = opts.loop;
+    }
+
+    this.direction = 'normal';
+    if (opts.direction !== undefined) {
+      this.direction = opts.direction;
+    }
+
+    this.autoplay = true;
+    if (opts.autoplay !== undefined) {
+      this.autoplay = opts.autoplay;
+    }
+
+    this.duration = 1000;
+    if (opts.duration !== undefined) {
+      this.duration = opts.duration;
+    }
+
+    // callbacks
+    this.update = opts.update;
+    this.start = opts.start;
+    this.run = opts.run;
+    this.end = opts.end;
+
+    // create animations
+    this.tracks = _createTracks(targets, props, opts);
+    this.duration = findMaxDuration(this.tracks);
+
+    this.paused = true;
+    this.began = false;
+    this.completed = false;
+    this.reversed = false;
+
+    this.currentTime = 0;
+    this.remaining = 0;
+    this.now = 0;
+    this.startTime = 0;
+    this.lastTime = 0;
+
+    // children
+    this.children = [];
+
+    this.reset();
+
+    if (this.autoplay) {
+      this.play();
+    }
+  }
+
+  adjustTime(time) {
+    return this.reversed ? this.duration - time : time;
+  }
+
+  toggle() {
+    this.reversed = !this.reversed;
+  }
+
+  countIteration() {
+    if (this.remaining && this.remaining !== true) {
+      this.remaining--;
+    }
+  }
+
+  invokeCallback(cb) {
+    if (this[cb]) {
+      this[cb](this);
+    }
+  }
+
+  syncChildren(time) {
+    let children = this.children;
+    let childrenLength = children.length;
+    
+    if (time >= this.currentTime) {
+      for (let i = 0; i < childrenLength; i++) {
+        children[i].seek(time);
+      }
+    } else {
+      for (let i = childrenLength; i--;) {
+        children[i].seek(time);
+      }
+    }
+  }
+
+  _setTracksProgress(time) {
+    for (let i = 0; i < this.tracks.length; i++) {
+      const track = this.tracks[i];
+      for (let j = 0; j < track.keys.length; j++) {
+        const key = track.keys[j];
+        if (key.startTime <= time && key.endTime >= time) {
+          const elapsed = clamp(time - key.startTime - key.delay, 0, key.duration) / key.duration;
+          const eased = isNaN(elapsed) ? 1 : key.easing(elapsed);
+
+          track.lerp(track.target[track.property], key.from, key.to, eased);
```

Comment:
track.target[track.property] = track.lerp(track.target[track.property], key.from, key.to, eased);